### PR TITLE
Improved hover support: Structs, qualified calls and types, more info for modules

### LINKED ISSUE #348: Support `textDocument/hover` for additional entities
#331 implements the strategy for resolving entities under the cursor and providing hover docs, but only supports modules. Most important additions that I can think of:

### `{:module, module}`

Implemented in #331. 

### `{:struct, module}`

When a module is being used as a struct, hover docs should highlight the struct fields, defaults, and available type information. Type information can be retrieved by fetching the types for a module and looking for those that define a type for the struct at the top level. 

### `{:call, module, fun, arity}`

Function or macro calls. Should resolve the module for local/import/Kernel calls. 

Arity should be best-effort. If the current arity used in the document matches a defined arity for the function, that doc string should be shown first. I think docs for all arities should be visible by scrolling, though.

### `{:type, module, fun, arity}`

Calls in type or spec declarations. Should resolve the module for local/import/Kernel types. 

---

Some additional things that would be cool to support but need to be thought through a bit more:

- Params used in a function body. Could potentially show what contextual information we have for that var, perhaps by showing the signature of the current function with that parameter highlighted.
- Module attribute usage. A basic implementation could show the closest static definition. This gets more complex with quoting, however.

### PR DESCRIPTION
**Pull request by zachallaun**
This PR builds on #331, adding hover support for additional entities:

- Structs
- Qualified calls
- Qualified types

Improvements:

- Generally improved formatting for all hover content.
- Respond with range information to support editor highlighting.
- Include callback signatures and docs for modules that are behaviours.

Notes:

- Add vendored `Future.Code.Typespec`, which includes required fixes in `@spec` to make Dialyzer happy.
- Update `:sourceror` dep to the latest commit until a new release is cut. This latest version updates Sourceror's zippers to be a `%Zipper{}` struct and includes fixes/improvements to `Sourceror.get_range/1`.
- `Lexical.Ast` now uses the `unescape: false` option when parsing code. See ba4223df5c4020c3dc6bc104b73c208496837cb7 for more context.

### CODE REVIEW

**Comment by zachallaun** [comment]
Struct hover:

image

Qualified call hover:

image

**Comment by zachallaun** [comment]
@scohen would appreciate your input on this:

In d1edb31da51485a183ac173fd274b8deeac27a2e, I added some additional utilities to `Lexical.RemoteControl.Modules` that deal with extracting information from BEAM object code -- docs, types, specs, and callbacks. While I was looking at `Code.Typespec` to better understand some things, I noticed that all of the functions will also accept the beam object binary, which means they don't have to read it off disk (and we don't have to wait for just-compiled files to be _written_ to disk).

Unfortunately, this introduced a Dialyzer error -- the `@spec` for these functions only accepts `module()` even though the guards are for `is_atom(module) or is_binary(module)`. [See here](https://github.com/elixir-lang/elixir/blob/v1.15.5/lib/elixir/lib/code/typespec.ex#L124).

Options I can think of:

1. Revert the changes and just operate on modules
2. Vendor `Code.Typespec` into `Future.Code.Typespec`, which I think would cause the Dialyzer error to be ignored because `lib/elixir/lib/future` is in `dialyzer.ignore-warnings`
3. Add a bunch of `@dialyzer {:nowarn_function, ...}` attributes, but they'd have to go both in the `Modules` utility and where it's being used, which is gross.

**Comment by scohen** [comment]
@zachallaun this looks like a typespec bug in elixir, we should fix that too, then pull the fix into `Future.Code`, which we do already.

File: apps/common/lib/lexical/ast.ex [hunk 1]
```diff
@@ -87,6 +87,18 @@ defmodule Lexical.Ast do
...
 
+  @doc """
+  Guard used to determined whether an AST node is a call.
+
+  ## Example
+
+      def my_fun({call, _meta, args}) when is_call(call, args), do: ...
+
+  """
+  @non_call_nodes [:., :__aliases__, :__block__, :"::", :{}, :|>, :%, :%{}]
```
Comment (scottming) [inline apps/common/lib/lexical/ast.ex:98]: I'm curious where you got this list from?

**Review by scottming — COMMENTED** [review]
For me, it looks good overall. However, I'm not sure if we need the `assert_hover` abstraction. Can you explain the reason behind it?

File: apps/common/lib/lexical/ast.ex (hunk 1 above)
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex]: I got the idea from ElixirSense, though ElixirSense doesn't contain `:__block__` which I think is incorrect. That said, I'd actually like for this (and others) to live in `Sourceror` (see [this issue](https://github.com/doorgan/sourceror/issues/105)), so it's possible that this will be removed prior to merging.

File: apps/server/test/lexical/server/code_intelligence/entity_test.exs [hunk 2]
```diff
@@ -480,6 +480,164 @@ defmodule Lexical.Server.CodeIntelligence.EntityTest do
...
+    test "qualified call nested in a pipe", %{project: project} do
+      code = ~q[
+        1
+        |> other()
+        |> MyModule.|some_function(2, 3)
+        |> other()
+      ]
+
+      assert {:ok, {:call, MyModule, :some_function, 3}, resolved_range} = resolve(project, code)
+      assert resolved_range =~ ~S[|> «MyModule.some_function»(2, 3)]
```
Comment (scottming) [inline apps/server/test/lexical/server/code_intelligence/entity_test.exs:609]: I think we do not need to highlight the `module` range,`MyModule.«some_function»` is more precise

**Comment by zachallaun** [comment]
@scohen @scottming This PR is ready for review. Please see the updated OP for additional context/notes.

While I was originally planning for this PR to also include hover support for unqualified calls, I chose to punt on that until the next one. Resolving unqualified calls is a lot more complex, since it requires import tracking, so I plan to address it separately.

File: apps/server/test/lexical/server/code_intelligence/entity_test.exs [hunk 3]
```diff
@@ -480,6 +480,215 @@ defmodule Lexical.Server.CodeIntelligence.EntityTest do
     end
   end
 
+  describe "struct resolve/2" do
```
[the remark this answers is not in the document]
Comment (scottming) [reply apps/server/test/lexical/server/code_intelligence/entity_test.exs:483]: Another point I think we can improve is showing default values for structs in the struct section.

Of course, addressing this in a future PR is fine.

**Review by scottming — APPROVED** [review]
Only a few minor comments were left. Overall, I feel that this improvement is excellent.

**Comment by scottming** [comment]
Another distracting element that I think can be removed is the block separator symbol `---`. I find it too bright in the nvim editor, which causes distraction. 

Of course, it would be good to have the option to customize the configuration and decide whether we need this separator.

CleanShot 2023-09-18 at 19 30 13@2x

**Comment by zachallaun** [comment]
@scottming I think that's fair; perhaps the `---` can be reserved for only separating between arities. (It shows docs for arities >= the one currently used.)

This is more of a meta-thought, but I wish there were a way for all of us to easily test these kinds of features in all major editors. Maybe it's worth setting up an "editor test configuration" repo that contains config for all the major editors and scripts to launch the editor using that config. Needs more thought, but could be useful in the long term.

File: apps/common/lib/lexical/ast.ex [hunk 4]
```diff
@@ -267,21 +267,73 @@ defmodule Lexical.Ast do
...
+  @spec path_at(Document.t(), Position.t()) ::
+          {:ok, [Macro.t(), ...]} | {:error, :not_found | parse_error()}
+  def path_at(%Document{} = document, %Position{} = position) do
+    with {:ok, ast} <- from(document) do
+      path_at(ast, position)
+    end
+  end
+
+  @spec path_at(Macro.t(), Position.t()) :: {:ok, [Macro.t(), ...]} | {:error, :not_found}
+  def path_at(ast, %Position{} = position) do
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex:289]: Yes! I meant to add tests for this; thank you for the reminder.

File: apps/server/lib/lexical/server/code_intelligence/entity.ex [hunk 5]
```diff
@@ -10,34 +10,37 @@ defmodule Lexical.Server.CodeIntelligence.Entity do
...
   """
   @spec resolve(Document.t(), Position.t()) :: {:ok, resolved, Range.t()} | {:error, term()}
   def resolve(%Document{} = document, %Position{} = position) do
-    with {:ok, %{context: context, begin: begin_pos, end: end_pos}} <-
-           Ast.surround_context(document, position),
-         {:ok, resolved, {begin_pos, end_pos}} <-
-           resolve(context, {begin_pos, end_pos}, document, position) do
+    with {:ok, surround_context} <- Ast.surround_context(document, position),
+         {:ok, resolved, {begin_pos, end_pos}} <- resolve(surround_context, document, position) do
+      Logger.info("Resolved entity: #{inspect(resolved)}")
```
[the remark this answers is not in the document]
Comment (scottming) [reply apps/server/lib/lexical/server/code_intelligence/entity.ex:32]: I do not have a strong opinion here.

Given that this feature is still relatively new and will continue to evolve in the near future, perhaps we can keep it for a while.

File: apps/server/lib/lexical/server/code_intelligence/entity.ex [hunk 6]
```diff
@@ -50,55 +53,141 @@ defmodule Lexical.Server.CodeIntelligence.Entity do
...
+  end
+
+  defp expand_alias(_, _document, _position), do: :error
+
+  # Pipes:
+  defp arity_at_position([{:|>, _, _} = pipe | _], position) do
+    {_call, _, args} =
+      pipe
+      |> Macro.unpipe()
+      |> Stream.map(fn {ast, _arg_position} -> ast end)
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/server/lib/lexical/server/code_intelligence/entity.ex:151]: Yes, this prevents an additional list from being created, since this returns a lazy enumerable that will only yield a mapped value when "requested" during the `Enum.find`. 

That said, I should actually be using [`Enum.find_value`](https://hexdocs.pm/elixir/1.15.5/Enum.html#find_value/3) here instead.

File: apps/server/lib/lexical/server/provider/handlers/hover.ex [hunk 7]
```diff
@@ -24,21 +30,187 @@ defmodule Lexical.Server.Provider.Handlers.Hover do
...
+       when fn_or_macro in [:function, :macro] do
+    with {:ok, call_header} <- call_header(entry) do
+      specs = Enum.map_join(entry.defs, "\n", &("@spec " <> &1))
+
+      [
+        call_header,
+        if specs != "" do
+          specs
+          |> Markdown.code_block()
+          |> Markdown.section(header: "Specs")
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/server/lib/lexical/server/provider/handlers/hover.ex:135]: It's largely to fix a rendering issue in VS Code hover when you have a `---` followed immediately be a code block. Since getting rid of the horizontal rules though, I'll change this too.

File: apps/server/test/lexical/server/code_intelligence/entity_test.exs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/server/test/lexical/server/code_intelligence/entity_test.exs:483]: > Another point I think we can improve is showing default values for structs in the struct section.

I'd like to do this! I'm thinking separate PR though, because I think the nicest way to do this is to merge the defaults and the type spec, so you see something like:

```elixir
%MyModule{
  foo: :default_foo :: atom(),
  bar: "baz" :: MyModule.custom_type()
}
```

and since that's non-trivial and has some edge cases, IMO best in another PR.

[the remark this answers is not in the document]
Comment (scottming) [reply apps/server/test/lexical/server/code_intelligence/entity_test.exs:483]: or, format like this: 

```elixir
%MyModule{
  foo: atom() \\ :default_foo,
  bar: MyModule.custom_type() \\ "baz"
}
```

Doing this formatting should have some work, feel free to do this at another PR

File: projects/lexical_test/lib/lexical/test/cursor_support.ex [hunk 8]
```diff
@@ -1,49 +1,98 @@
...
+      iex> pop_cursor(code, document: "my_doc.ex")
+      {
+        %Position{line: 2, column: 12},
+        %Document{uri: "file:///my_doc.ex", ...}
+      }
+
+  """
+  @spec pop_cursor(text :: String.t(), [opt]) :: {Position.t(), String.t() | Document.t()}
+        when opt: {:cursor, String.t()} | {:as, :text | :document} | {:document, String.t()}
+  def pop_cursor(text, opts \\ []) do
```
Comment (scottming) [inline projects/lexical_test/lib/lexical/test/cursor_support.ex:64]: I'm like this change; `pop_cursor` returns `{cursor, text_without_cursor}` really makes sense.

File: apps/common/test/lexical/ast_test.exs [hunk 9]
```diff
@@ -47,20 +47,77 @@ defmodule Lexical.AstTest do
...
+      ]
+
+      assert {:ok,
+              [
+                {:bar, _, nil},
+                {:def, _, _},
+                {_, _},
+                [{_, _}],
+                {:defmodule, _, _}
+              ]} = path_at(code)
```
Comment (scottming) [inline apps/common/test/lexical/ast_test.exs:108]: So this function actually replaces the `__CURSOR__` of the `cursor_path` function's result with the exact token?

**Review by scottming — APPROVED** [review]
Very good, you're awesome, @zachallaun . Thank you.

Another improvement may be to convert the `Examples` section to an Elixir code block, but feel free to do that in another PR.

CleanShot 2023-09-19 at 22 07 10@2x

File: apps/common/test/lexical/ast_test.exs (hunk 9 above)
Comment (zachallaun) [reply apps/common/test/lexical/ast_test.exs:108]: Sort of, I think. The goal of `path_at` is to get the path to the innermost terminal node at a given position, where terminal node is essentially a node that doesn't have any (meaningful) children. To do this, `path_at` uses the entire AST and can give a more accurate path.

To illustrate the difference, I played around in `MIX_ENV=test iex -S mix` a bit:

```elixir
iex(4)> {pos, doc} = Lexical.Test.CursorSupport.pop_cursor(~q[
...(4)> defmodule MyModule do
...(4)>   def foo(bar) do
...(4)>     bar         
...(4)>     |> baz()
...(4)>     |> Other.|buzz()
...(4)>     |> Other.more()
...(4)>   end
...(4)> end
...(4)> ], as: :document)

iex(5)> Lexical.Ast.path_at(doc, pos)
{:ok,
 [
   {:., [line: 5, column: 13],
    [
      {:__aliases__, [last: [line: 5, column: 8], line: 5, column: 8], [:Other]},
      :buzz
    ]},
   {{:., [line: 5, column: 13],
     [{:__aliases__, [last: [line: 5, ...], line: 5, ...], [:Other]}, :buzz]},
    [closing: [line: 5, column: 19], line: 5, column: 14], []},
   {:|>, [newlines: 1, line: 5, column: 5],
    [
      {:|>, [newlines: 1, line: 4, ...], [{...}, ...]},
      {{:., [...], ...}, [closing: [...], ...], []}
    ]},
   {:|>, [newlines: 1, line: 6, column: 5],
    [{:|>, [newlines: 1, ...], [...]}, {{:., ...}, [...], ...}]},
   {{:__block__, [line: 2, column: 16], [:do]},
    {:|>, [newlines: 1, line: 6, ...], [{...}, ...]}},
   [{{:__block__, [line: 2, ...], [...]}, {:|>, [...], ...}}],
   {:def, [do: [line: 2, ...], end: [...], ...], [{...}, ...]},
   {{:__block__, [...], ...}, {:def, ...}},
   [{{...}, ...}],
   {:defmodule, ...}
 ]}

iex(6)> Lexical.Ast.cursor_path(doc, pos)  
[
  {:__cursor__, [closing: [line: 5, column: 25], line: 5, column: 14], []},
  {:|>, [newlines: 1, line: 5, column: 5],
   [
     {:|>, [newlines: 1, line: 4, ...], [{...}, ...]},
     {:__cursor__, [closing: [...], ...], []}
   ]},
  {{:__block__, [line: 2, column: 16], [:do]},
   {:|>, [newlines: 1, line: 5, column: 5], [{:|>, ...}, {...}]}},
  [
    {{:__block__, [line: 2, column: 16], [:do]},
     {:|>, [newlines: 1, ...], [...]}}
  ],
  {:def, [do: [line: 2, column: 16], end: [line: 5, ...], line: 2, ...],
   [{:foo, ...}, [...]]},
  {{:__block__, [line: 1, ...], [...]}, {:def, [...], ...}},
  [{{:__block__, ...}, {...}}],
  {:defmodule, [...], ...}
]
```

Since `cursor_path` only operates on the fragment before the cursor, it doesn't give any contextual information about the call that the cursor is in.

**Comment by zachallaun** [comment]
> Another improvement may be to convert the Examples section to an Elixir code block, but feel free to do that in another PR.

@scottming This is a great idea! I wanted to see if I could sneak it in, but it's unfortunately a bit more complicated than it first seems and, I think, should be in its own PR. Basically, we need to handle all of these cases:

Indented code in an explicit code block shouldn't be re-wrapped:

    ```
        some_code
    ```
    
Multiple indented blocks separated by empty lines should be joined together into a single block:

    ## Examples

        some_code

        with_an_empty_line_between

Code indented relative to a parent element (like a list item) should be wrapped, but retain the parent's indentation:

    ## List of items

      - Something with a code example:

            some_code_4_spaces_in_from_Something

I think there's a somewhat simple way to do this that doesn't require a full Markdown parser by basically map-reducing over the lines, looking for lines that start with <code>```</code> to ignore until the block ends, and then tracking the indentation level of the previous line so that indented blocks can be identified by `current_indent == previous_line_indent + 4`.

But it's enough work that I don't want to include it here :)

File: apps/server/lib/lexical/server/provider/markdown.ex [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+  alias Lexical.Protocol.Types.Markup
+
+  @type markdown :: String.t()
+
+  @doc """
+  Converts a string of Markdown into LSP markup content.
+  """
+  @spec to_content(markdown) :: Markup.Content.t()
+  def to_content(markdown) when is_binary(markdown) do
```
Comment (scohen) [inline apps/server/lib/lexical/server/provider/markdown.ex:14]: I don't know if there is where to comment, but I've noticed that the markdown comment has a bunch of newlines at the end of it --enough to scroll the content off the limited area that emacs displays it.
You _do_ have to scroll to trigger this behavior, I thought I'd mention it.

Here's an example of it scrolled.

Screenshot 2023-09-19 at 9 05 35 AM

File: apps/server/lib/lexical/server/provider/markdown.ex [hunk 11]
```diff
@@ -0,0 +1,76 @@
+defmodule Lexical.Server.Provider.Markdown do
```
Comment (scohen) [inline apps/server/lib/lexical/server/provider/markdown.ex:1]: I really like this abstraction.

File: apps/server/lib/lexical/server/provider/handlers/hover.ex [hunk 12]
```diff
@@ -24,21 +26,171 @@ defmodule Lexical.Server.Provider.Handlers.Hover do
...
+  defp hover_content({:call, module, fun, arity}, env) do
+    with {:ok, %Docs{} = module_docs} <- RemoteControl.Api.docs(env.project, module),
+         {:ok, entries} <- Map.fetch(module_docs.functions_and_macros, fun) do
+      sections =
+        entries
+        |> Enum.sort_by(& &1.arity)
+        |> Enum.filter(&(&1.arity >= arity))
+        |> Enum.map(&entry_content/1)
+
+      {:ok, Markdown.join_sections(sections, "\n\n---\n\n")}
```
Comment (scohen) [inline apps/server/lib/lexical/server/provider/handlers/hover.ex:61]: Should `\n\n---\n\n` be something in markdown, like `Markdown.separator`?

File: apps/server/test/lexical/server/code_intelligence/entity_test.exs [hunk 13]
```diff
@@ -480,30 +478,240 @@ defmodule Lexical.Server.CodeIntelligence.EntityTest do
...
+    test "expands current module", %{project: project} do
+      code = ~q[
+            defmodule Example do
+              %|__MODULE__{}
+            end
+          ]
+
+      assert {:ok, {:struct, Example}, resolved_range} = resolve(project, code)
+      assert resolved_range =~ ~S[  %«__MODULE__»{}]
+    end
```
Comment (scohen) [inline apps/server/test/lexical/server/code_intelligence/entity_test.exs:539]: another odd case would be `__MODULE__.Substruct|{}`, which has occurred in the past.

how about cases with aliased modules 
```elixir
`q[
  alias Something.Else 
  %Else.Another|{}
 ]
```
should resolve `Something.Else.Another`

File: apps/server/test/lexical/server/code_intelligence/entity_test.exs [hunk 14]
```diff
@@ -480,30 +478,240 @@ defmodule Lexical.Server.CodeIntelligence.EntityTest do
...
+  end
+
+  describe "call resolve/2" do
```
Comment (scohen) [inline apps/server/test/lexical/server/code_intelligence/entity_test.exs:542]: same comment about aliases here. Given how thorough these tests are, maybe resolve isn't supposed to resolve aliases?

**Review by scohen — COMMENTED** [review]
This looks very good. I like the abstractions and what you've done with cursor support. 

I have some comments along the lines of how positions are represented and if we can do better in tests with defining ranges and positions, but this is nearly ready.

File: apps/common/lib/lexical/ast.ex [hunk 15]
```diff
@@ -187,9 +187,40 @@ defmodule Lexical.Ast do
...
+          {:ok, [Macro.t(), ...]} | {:error, :not_found}
+  def path_at(%Document{} = document, %Position{} = position) do
+    with {:ok, ast} <- from(document) do
+      path_at(ast, position)
+    end
+  end
+
+  def path_at(ast, %Position{} = position) do
+    path =
+      Future.Macro.path(ast, fn node ->
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex:207]: I thought the same thing at first, but our `Future.Macro` isn't a fully vendored `Macro`, it only contains the new `path` API.

[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex:207]: (Which means we can't alias it because we need the non-vendored `Macro` for other things.)

File: apps/common/mix.exs [hunk 16]
```diff
@@ -33,7 +33,7 @@ defmodule Common.MixProject do
   defp deps do
     [
       {:lexical_shared, path: "../../projects/lexical_shared"},
-      {:sourceror, "~> 0.12"},
+      {:sourceror, github: "doorgan/sourceror", ref: "b0ec3cc"},
```
[the remark this answers is not in the document]
Comment (doorgan) [reply apps/common/mix.exs:36]: v0.14.0 just dropped! Definitely feel free to ping me whenever you need me to cut a new release

File: apps/common/lib/lexical/ast.ex (hunk 15 above)
[the remark this answers is not in the document]
Comment (scohen) [reply apps/common/lib/lexical/ast.ex:207]: really? Why?

File: apps/common/test/lexical/ast_test.exs [hunk 17]
```diff
@@ -113,4 +170,98 @@ defmodule Lexical.AstTest do
...
+
+  describe "contains_position?/2 single line node" do
+    setup do
+      [single_line_ast] = ast(~q|
+        [
+          single_line_call(1, 2, 3),
+        ]
+      |)
+
+      range = {{2, 3}, {2, 27}}
```
[the remark this answers is not in the document]
Comment (scohen) [reply apps/common/test/lexical/ast_test.exs:182]: or `ast_with_range` ?

File: apps/common/lib/lexical/ast.ex (hunk 15 above)
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex:207]: > really? Why?

Do you mean why isn't all of `Macro` vendored or why can't we alias it? or something else?

File: apps/server/test/lexical/server/code_intelligence/entity_test.exs (hunk 14 above)
Comment (zachallaun) [reply apps/server/test/lexical/server/code_intelligence/entity_test.exs:542]: It does (or should) resolve aliases. Good call, will add a test for this and aliased structs as suggested above.

File: apps/common/lib/lexical/ast.ex [hunk 18]
```diff
@@ -187,9 +187,40 @@ defmodule Lexical.Ast do
...
+        terminal_node?(node) and contains_position?(node, position)
+      end)
+
+    case path do
+      nil -> {:error, :not_found}
+      path -> {:ok, path}
+    end
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex:214]: The reason I used `{:error, :not_found}` is because it's also possible that this function returns a syntax error from parsing, which is also a 2-element `:error` tuple, and it seems a little icky for a function to return both `:error` and `{:error, something}`.

File: apps/server/lib/lexical/server/provider/markdown.ex (hunk 10 above)
Comment (zachallaun) [reply apps/server/lib/lexical/server/provider/markdown.ex:14]: This is super weird and I can't recreate it in a test. I'm thinking it might be a bug in the Emacs client? E.g. when I hover over `RemoteControl`, it's not scrollable:

image

File: apps/remote_control/lib/lexical/remote_control/modules.ex [hunk 19]
```diff
@@ -47,8 +44,135 @@ defmodule Lexical.RemoteControl.Modules do
...
+          {:ok, beam :: binary()} | {:error, reason}
+        when reason:
+               :embedded
+               | :badfile
+               | :nofile
+               | :on_load_failure
+               | :unavailable
+               | :get_object_code_failed
+  def ensure_beam(module) when is_atom(module) do
+    with {:module, _} <- Code.ensure_compiled(module),
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/remote_control/lib/lexical/remote_control/modules.ex:75]: The idea is that this allows doc fetching to operate independently of compilation. I believe it was originally needed because there were situations in tests where a newly introduced module was loaded but not compiled, but I've since refactored the hover tests and this may not be needed anymore.

File: apps/common/lib/lexical/ast.ex (hunk 15 above)
[the remark this answers is not in the document]
Comment (scohen) [reply apps/common/lib/lexical/ast.ex:207]: Why we can't alias all calls to `Macro` with `Future.Macro`.

File: apps/common/lib/lexical/ast.ex (hunk 18 above)
[the remark this answers is not in the document]
Comment (scohen) [reply apps/common/lib/lexical/ast.ex:214]: i see.

File: apps/server/lib/lexical/server/provider/markdown.ex (hunk 10 above)
Comment (scohen) [reply apps/server/lib/lexical/server/provider/markdown.ex:14]: a bug in `lsp_node`, no way. 🤣

File: apps/common/lib/lexical/ast.ex (hunk 15 above)
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex:207]: Because `Future.Macro` only contains `path/2`, so any call to `Macro.not_path/2` fails.

[the remark this answers is not in the document]
Comment (scottming) [reply apps/common/lib/lexical/ast.ex:207]: Yes, because Macro has dependencies on many other Elixir erlang modules, and I didn't want to copy these extra Erlang modules to Lexical at the time.

[the remark this answers is not in the document]
Comment (scohen) [reply apps/common/lib/lexical/ast.ex:207]: that kind of defeats the point of the strategy here, no? 
The idea was to use the future stuff.

[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/common/lib/lexical/ast.ex:207]: I don't think it defeats the point. We're still using the new stuff where it matters (the `path` API introduced in 1.14).

If we need anything else new from `Macro`, we can vendor in the whole thing, but I def don't think it needs to happen in this particular PR.

[the remark this answers is not in the document]
Comment (scottming) [reply apps/common/lib/lexical/ast.ex:207]: @zachallaun Yes, I had the same thoughts

**Comment by zachallaun** [comment]
Draft squash commit:

* Include the range of resolved entities when responding to hover requests

This allows editors to highlight the token that the hover content applies to. For instance, if you hovered `Baz` in the module `Foo.Bar.Baz`, the entire module will be highlighted, not just the hovered segment in the module.

* Improve formatting and docs when hovering modules

Behaviours (modules that have `@callback` attributes) will now show callback specs and below any module docs.

* Don't show return content when hovering modules without docs

This applies to modules that use `@moduledoc false` or modules that are missing a `@moduledoc` attribute and don't have other additional docs to show (e.g. behaviour callbacks).

* Add support for hovering structs

When hovering a module being used as a struct, e.g. `%MyStruct{}`,  `MyStruct.t` types will now be shown before any additional module docs.

* Add support for hovering qualified functions, macros, and types

Qualified calls are those that begin with a module, e.g. `MyModule.my_fun()`. When hovering qualified functions or macros, the signatures of those calls will be shown followed by any specs and docs. When hovering qualified types, the type definition will be shown followed by any docs.

For functions with multiple arities, the arity currently being used will try to be inferred and those definitions/docs shown first, followed by those for any higher arities. 

* [Internal] Add additional utilities to `Lexical.Ast`, including `path_at/2` and `contains_position?/2`

* [Internal] Add `Lexical.Test.CursorSupport.pop_range/2`

* [Internal] Add `Lexical.Test.RangeSupport.pop_range/1`

* [Internal] Update Sourceror to 0.14

This version bump includes more complete and resilient APIs for getting ranges, as well as a new `%Zipper{}` struct that replaces the old `{tree, meta}` tuple that previously was used to represent zippers.

**Review by scohen — APPROVED** [review]
